source-code/core2/src/messages/query.ts

```diff
@@ -1,15 +1,48 @@
 import type { Message } from "./schema.js"
+const inlang: any = {}
 
-// TODO
-// 1. Query builder?
-// 2. Filtering etc how?
-//
-// inspiration: Prisma CRUD https://www.prisma.io/docs/concepts/components/prisma-client/crud
-export type MessagesQueryApi_1 = {
-	create: (message: Message) => void
-	get: () => Message[]
-	update: () => void
-	delete: () => void
+/* --------------------- TYPES --------------------- */
+
+type UniqueFilterType = {
+	id: string
+	languageTag: string
+	// more properties in the future
+}
+
+type QueryOptions = {
+	where?: Partial<UniqueFilterType>[]
 }
 
-// query.
+export type MessagesQuery = {
+	get: (args: QueryOptions) => Message[]
+	update: (data: Message[]) => Message[]
+	list: () => Message[]
+}
+
+/* --------------------- CRUD --------------------- */
+
+// - create
+const message = inlang.messages.update([
+	{ id: "myMessageId", languageTag: "en", pattern: [{ type: "Text", value: "Hello World" }] },
+]) as ReturnType<MessagesQuery["update"]>
+
+// - get
+const messageOrMessages = inlang.messages.get({
+	where: { id: "myMessageId", languageTag: "en" },
+}) as ReturnType<MessagesQuery["get"]>
+
+// - update
+const messageOrMessagesUpdated = inlang.messages.update([
+	{ id: "myMessageId", languageTag: "en", pattern: [{ type: "Text", value: "Hello World" }] },
+]) as ReturnType<MessagesQuery["update"]>
+
+// - delete
```

Comment:
I don't think that update to an undefined pattern is the same as a dedicated delete. This leads to confusion.

Reply:
I think having to decide and flesh out the cases between 5+ crud functions in apps are the bigger burden as passing `undefined` to one function and let core handle the rest.